Use nix shell

yakimant opened the pull request:
Fixes https://github.com/status-im/status-app/issues/16281

### What does the PR do

1. Swiches to Nix flakes to run Nix shell faster (`nix develop`)
2. Adds `direnv` setup (`nix-direnv)

### How to test

- Setup local tools: nix, direnv
- enter shell by (one of the ways):
  - `direnv allow`
  - `nix-shell`
- run the usual Linux targets

### TODO

- [x] Issues running an app with `export FLAG_DAPPS_ENABLED=1`
- [x] `nix develop` issue: `unsupported tarball input attribute 'lastModified'`. Maybe wrong `nix` version
- [x] make sure `nix.conf` is loaded in each case (direnv, develop, shell)
- [x] `make run` fails to run in nix shell
- [x] Appimage takes QtWebEngineProccess from /nix/store if present, it is not patched, so breaks
- [ ] rebase from master
- [ ] extract AppRun, init_app_dir, fix_app_dir to separate PR (AppImage related)
- [ ] Remove `flake-compat`, we dont need `nix-shell`
- [ ] Update nixpkgs to stable, remove unnessasary overrides, add unstable nixpkgs for Qt if needed
- [ ] Research on nixGL vs nix-gl-host vs nix-system-graphics solution for OpenGL problem
- [ ] Test Makefile dev targets
- [ ] Test running the app on Wayland
- [ ] Opt-out of nix shell (using system libs/tools for both building and running) - see `check-nix-shell`
- [ ] Document to setup environment - `nix/README.md` https://github.com/status-im/status-app/pull/16359#issuecomment-2379237987

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | f0738529 | [#1](https://ci.status.im/job/status-desktop/job/prs/job/linux/job/x86_64/job/package-nix/job/PR-16359/1/) | 2024-09-18 14:54:30 | ~2 min | `linux-nix/x86_64` | [:page_facing_up:`log`](https://ci.status.im/job/status-desktop/job/prs/job/linux/job/x86_64/job/package-nix/job/PR-16359/1/consoleText) |

status-im-auto commented:
#### :heavy_check_mark: [status-desktop/prs/windows/x86_64/package/PR-16359#1](https://ci.status.im/job/status-desktop/job/prs/job/windows/job/x86_64/job/package/job/PR-16359/1/) :small_blue_diamond: ~26 min :small_blue_diamond: f0738529 :small_blue_diamond: [:package: windows/x86_64 package](https://status-im-desktop-prs.ams3.digitaloceanspaces.com/StatusIm-Desktop-240918-145155-f07385-pr16359-x86_64.exe)

status-im-auto commented:
#### :heavy_check_mark: [status-desktop/prs/windows/x86_64/package/PR-16359#3](https://ci.status.im/job/status-desktop/job/prs/job/windows/job/x86_64/job/package/job/PR-16359/3/) :small_blue_diamond: ~23 min :small_blue_diamond: 3a21cc33 :small_blue_diamond: [:package: windows/x86_64 package](https://status-im-desktop-prs.ams3.digitaloceanspaces.com/StatusIm-Desktop-240919-105343-3a21cc-pr16359-x86_64.exe)

status-im-auto commented:
#### :heavy_check_mark: [status-desktop/prs/windows/x86_64/package/PR-16359#4](https://ci.status.im/job/status-desktop/job/prs/job/windows/job/x86_64/job/package/job/PR-16359/4/) :small_blue_diamond: ~28 min :small_blue_diamond: 240b4361 :small_blue_diamond: [:package: windows/x86_64 package](https://status-im-desktop-prs.ams3.digitaloceanspaces.com/StatusIm-Desktop-240919-124406-240b43-pr16359-x86_64.exe)

yakimant commented:
@igor-sirotin, @alexjba, @caybro, @micieslak,
guys, I need feedback from developers about this approach.

This PR is enforcing Nix environment for Linux builds.
You will need to have Nix and direnv installed.

As a benefit you will have:
- Fully reproducable and managed dev environment (no need to install system tools, no different versions)
- Newer Qt version - 5.15.8 now, v5.15.15 easily achivable

jakubgs commented on flake.nix:
```diff
@@ -0,0 +1,15 @@
+        overlays = [ (import ./nix/overlay.nix) ];
+      };
+    in {
+      devShells.x86_64-linux.default = import ./nix/shell.nix { inherit pkgs; };
```
This is probably a more kosher way.
```suggestion
      devShells.x86_64-linux.default = pkgs.callPackage ./nix/shell.nix { };
```

jakubgs reviewed (approved):
Not sure if I'm a fan of this `flake-compat` stuff, but other than that this looks good.

alexjba reviewed (approved):
Glad to see this change! 🍻

anastasiyaig commented:
@yakimant @siddarthkay we need to tell tests to use another artefact :)

status-im-auto commented:
#### :heavy_check_mark: [status-desktop/prs/windows/x86_64/package/PR-16359#5](https://ci.status.im/job/status-desktop/job/prs/job/windows/job/x86_64/job/package/job/PR-16359/5/) :small_blue_diamond: ~27 min :small_blue_diamond: 630d599d :small_blue_diamond: [:package: windows/x86_64 package](https://status-im-desktop-prs.ams3.digitaloceanspaces.com/StatusIm-Desktop-240925-083505-630d59-pr16359-x86_64.exe)

status-im-auto commented:
#### :heavy_check_mark: [status-desktop/prs/windows/x86_64/package/PR-16359#6](https://ci.status.im/job/status-desktop/job/prs/job/windows/job/x86_64/job/package/job/PR-16359/6/) :small_blue_diamond: ~25 min :small_blue_diamond: efb7ee44 :small_blue_diamond: [:package: windows/x86_64 package](https://status-im-desktop-prs.ams3.digitaloceanspaces.com/StatusIm-Desktop-240925-104008-efb7ee-pr16359-x86_64.exe)

status-im-auto commented:
#### :heavy_check_mark: [status-desktop/prs/windows/x86_64/package/PR-16359#8](https://ci.status.im/job/status-desktop/job/prs/job/windows/job/x86_64/job/package/job/PR-16359/8/) :small_blue_diamond: ~24 min :small_blue_diamond: 6e708fc8 :small_blue_diamond: [:package: windows/x86_64 package](https://status-im-desktop-prs.ams3.digitaloceanspaces.com/StatusIm-Desktop-240925-133122-6e708f-pr16359-x86_64.exe)

jakubgs commented on ci/Jenkinsfile.linux:
```diff
@@ -6,15 +6,7 @@ def isPRBuild = utils.isPRBuild()
-      args '--entrypoint="" --cap-add SYS_ADMIN --security-opt apparmor:unconfined --device /dev/fuse'
-    }
-  }
+  agent { label 'linux && x86_64 && nix-2.14' }
```
We are on `2.19` right now:
https://github.com/status-im/infra-ci/blob/67f9fc643a1e054d97c53a0e30446943d341cd98/ansible/roles/nix-setup/defaults/main.yml#L6

jakubgs commented on ci/Jenkinsfile.linux:
```diff
@@ -78,21 +67,21 @@ pipeline {
     stage('Package') {
       steps { script {
-        linux.bundle('tgz-linux')
+        linux.bundle('tgz-linux', 0, true)
```
```suggestion
        linux.bundle('tgz-linux', nixShell=true)
```

jakubgs reviewed (approved):
Looking good.

alexjba commented:
@yakimant could you please give me some instructions on how to build and run the app using the nix shell? I have a new linux environment with just a bare minimum of packages installed: build-essential, git, nix, direnv.

If I understand it correctly it's all I should have to get me started, right?

alexjba commented:
One other note: would it be a good chance to update the `ubuntu_build_setup.sh` and add a function that configures the new dependencies?

yakimant commented:
There are 2 options:
1. Setup direnv: `direnv allow`. Nix shell will start automatically every time you enter the work dir.
2. If you want to start the shell manualy: `nix develop`. If you forger make will fail with a notice to run a Nix shell.

In regards of `ubuntu_build_setup.sh` - yes, good idea. No so much will be left though (nix, direnv).

status-im-auto commented:
#### :heavy_check_mark: [status-desktop/prs/windows/x86_64/package/PR-16359#10](https://ci.status.im/job/status-desktop/job/prs/job/windows/job/x86_64/job/package/job/PR-16359/10/) :small_blue_diamond: ~21 min :small_blue_diamond: 195a1cc2 :small_blue_diamond: [:package: windows/x86_64 package](https://status-im-desktop-prs.ams3.digitaloceanspaces.com/StatusIm-Desktop-240926-112435-195a1c-pr16359-x86_64.exe)

status-im-auto commented:
#### :heavy_check_mark: [status-desktop/prs/windows/x86_64/package/PR-16359#12](https://ci.status.im/job/status-desktop/job/prs/job/windows/job/x86_64/job/package/job/PR-16359/12/) :small_blue_diamond: ~20 min :small_blue_diamond: 945b2f43 :small_blue_diamond: [:package: windows/x86_64 package](https://status-im-desktop-prs.ams3.digitaloceanspaces.com/StatusIm-Desktop-240926-143911-945b2f-pr16359-x86_64.exe)

alexjba commented:
@yakimant I think we'll need to find a way to share the nix packages. I've been naive enough to give it a spin in a linux VM and 15h later it's still compiling chromium.

I'm really a nix noob, but I see lots of resources on sharing the pre-compiled packages.
E.g. https://www.tweag.io/blog/2019-11-21-untrusted-ci/

yakimant commented:
We have it setup already:
https://github.com/status-im/status-desktop/commit/08d8c9ebf270d2b7fe65471c54dcfc001d2e9959

So you need to run
```
export NIX_USER_CONF_FILES := $(PWD)/nix/nix.conf
```
before entering the nix shell.

I will make sure it is set for all the cases, added to TODO.

status-im-auto commented:
#### :heavy_check_mark: [status-desktop/prs/windows/x86_64/package/PR-16359#13](https://ci.status.im/job/status-desktop/job/prs/job/windows/job/x86_64/job/package/job/PR-16359/13/) :small_blue_diamond: ~25 min :small_blue_diamond: 6e973dd3 :small_blue_diamond: [:package: windows/x86_64 package](https://status-im-desktop-prs.ams3.digitaloceanspaces.com/StatusIm-Desktop-240927-092650-6e973d-pr16359-x86_64.exe)

yakimant commented:
Setting up notes for future doc:

0) Install tools
```
sudo apt install git curl
./nix/script/setup.sh
```

Add to .zshrc:
```
. ~/.nix-profile/etc/profile.d/nix.sh
```

1) Run shell
First run can take a while to download the requirements

a) Manually
```
sudo apt install make
make shell
```

b) Automatically with direnv (every time you enter the folder)

```
sudo apt install direnv
```

Add to .zshrc:
```
eval "$(direnv hook zsh)"
export DIRENV_LOG_FORMAT='' # to be less verbose while entering folder
```

```
direnv allow
```

c) Using nix tools (unsupported):

```
export NIX_USER_CONF_FILES="${PWD}/nix/nix.conf"
nix develop # or nix-shell
```

status-im-auto commented:
#### :heavy_check_mark: [status-desktop/prs/windows/x86_64/package/PR-16359#15](https://ci.status.im/job/status-desktop/job/prs/job/windows/job/x86_64/job/package/job/PR-16359/15/) :small_blue_diamond: ~24 min :small_blue_diamond: 4458bc2e :small_blue_diamond: [:package: windows/x86_64 package](https://status-im-desktop-prs.ams3.digitaloceanspaces.com/StatusIm-Desktop-240927-125845-4458bc-pr16359-x86_64.exe)

alexjba commented:
> nix develop issue: unsupported tarball input attribute 'lastModified'. Maybe wrong nix version

Definitely wrong nix version. The build process can be started with the nix version installed by `status-go/nix/scripts/setup.sh`

yakimant commented:
`make run` running in nix shell fails:
```
INFO [10-01|12:48:05.205] Status backend initialized               backend=geth version=2.1.0-8-g2f71d9d9f commit=2f71d9d9f IpfsGatewayURL=https://ipfs.status.im/
qt.glx: qglx_findConfig: Failed to finding matching FBConfig for QSurfaceFormat(version 2.0, options QFlags<QSurfaceFormat::FormatOption>(), depthBufferSize -1, redBufferSize 1, greenBufferSize 1, blueBufferSize 1, alphaBufferSize -1, stencilBufferSize -1, samples -1, swapBehavior QSurfaceFormat::SingleBuffer, swapInterval 1, colorSpace QSurfaceFormat::DefaultColorSpace, profile  QSurfaceFormat::NoProfile)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig for QSurfaceFormat(version 2.0, options QFlags<QSurfaceFormat::FormatOption>(), depthBufferSize -1, redBufferSize 1, greenBufferSize 1, blueBufferSize 1, alphaBufferSize -1, stencilBufferSize -1, samples -1, swapBehavior QSurfaceFormat::SingleBuffer, swapInterval 1, colorSpace QSurfaceFormat::DefaultColorSpace, profile  QSurfaceFormat::NoProfile)
Could not initialize GLX
SIGABRT: Abnormal termination.
make[1]: *** [Makefile:866: run-linux] Aborted (core dumped)
```

yakimant commented:
https://github.com/nix-community/nixGL

status-im-auto commented:
#### :heavy_check_mark: [status-desktop/prs/windows/x86_64/package/PR-16359#16](https://ci.status.im/job/status-desktop/job/prs/job/windows/job/x86_64/job/package/job/PR-16359/16/) :small_blue_diamond: ~24 min :small_blue_diamond: 749feacd :small_blue_diamond: [:package: windows/x86_64 package](https://status-im-desktop-prs.ams3.digitaloceanspaces.com/StatusIm-Desktop-241002-100545-749fea-pr16359-x86_64.exe)

status-im-auto commented:
#### :heavy_check_mark: [status-desktop/prs/windows/x86_64/package/PR-16359#17](https://ci.status.im/job/status-desktop/job/prs/job/windows/job/x86_64/job/package/job/PR-16359/17/) :small_blue_diamond: ~23 min :small_blue_diamond: 3ce55948 :small_blue_diamond: [:package: windows/x86_64 package](https://status-im-desktop-prs.ams3.digitaloceanspaces.com/StatusIm-Desktop-241002-105443-3ce559-pr16359-x86_64.exe)

yakimant commented:
Nice list of Qt env vars:
https://github.com/NickHugi/PyKotor/blob/master/Tools/HolocronToolset/src/toolset/gui/widgets/settings/env_vars.py

yakimant commented:
Hm, impure shell with:
```
export LD_LIBRARY_PATH="/usr/lib/x86_64-linux-gnu:${LD_LIBRARY_PATH}"
export LD_LIBRARY_PATH="${LD_LIBRARY_PATH}:/usr/lib/x86_64-linux-gnu"
```

both break programs:
```
$ ls
Segmentation fault (core dumped)
```

yakimant commented:
I also catch this issue:
```
/nix/store/vfpyhifc47n4maql06phy9w2g0dmdsf3-qtwebengine-5.15.12/libexec/QtWebEngineProcess: error while loading shared libraries: libz.so.1: cannot open shared object file: No such file or directory
```

Which is weird:
- Why nix store path, not appimage?
- libz.so.1 is present on host

Maybe due to Nix installed on Ubuntu?

status-im-auto commented:
#### :heavy_check_mark: [status-desktop/prs/windows/x86_64/package/PR-16359#20](https://ci.status.im/job/status-desktop/job/prs/job/windows/job/x86_64/job/package/job/PR-16359/20/) :small_blue_diamond: ~24 min :small_blue_diamond: c56b36c3 :small_blue_diamond: [:package: windows/x86_64 package](https://status-im-desktop-prs.ams3.digitaloceanspaces.com/StatusIm-Desktop-241002-141018-c56b36-pr16359-x86_64.exe)

yakimant commented:
@alexjba, #4 is fixed, described here:
https://github.com/status-im/status-desktop/pull/16359#issuecomment-2388692765
Not sure about AppRun `LD_LIBRARY_PATH`, didn't think about it.

yakimant commented:
@anastasiyaig, issue with app crashing (or freezing) on `export FLAG_DAPPS_ENABLED=1` is workarounded by `QTWEBENGINE_CHROMIUM_FLAGS="--disable-gpu"`.

You can try the latest build to test.

Would be interesting to know if there are any issues with that.

alexjba commented:
> @alexjba, #4 is fixed, described here: [#16359 (comment)](https://github.com/status-im/status-desktop/pull/16359#issuecomment-2388692765) Not sure about AppRun `LD_LIBRARY_PATH`, didn't think about it.

Worth considering IMO. Looks like an overhead. All the best practices I could find on `LD_LIBRARY_PATH` specify that the best practice is not to use it 😆

alexjba commented:
> @anastasiyaig, issue with app crashing (or freezing) on `export FLAG_DAPPS_ENABLED=1` is workarounded by `QTWEBENGINE_CHROMIUM_FLAGS="--disable-gpu"`.
> 
> You can try the latest build to test.
> 
> Would be interesting to know if there are any issues with that.

Hopefully a more permanent fix here https://github.com/status-im/status-desktop/pull/16443

jakubgs commented:
Isn't it just the linux one?

anastasiyaig commented:
> Isn't it just the linux one?

there was a link to nix build job , i used that one before.. Maybe i need to use it again, hence asking :D

anastasiyaig commented:
@yakimant @alexjba tried a recent linux build from this PR with env var set as suggested

![image](https://github.com/user-attachments/assets/c4e95457-cade-4da7-8679-ffd381c5ca24)

yakimant commented:
@anastasiyaig, this PR is also getting rid of non-Nix build in Jenkins (subject to change).
So just use a Linux build.

alexjba commented:
> @anastasiyaig, very good! I'm not good in Stats / QtWebEngine functionality. Is everything alright with the broeser? We disabled hardware accelaration, so speed might be affected, some glitches appear, etc. (especially animations, font/images scaling/antialiasing,..)

Short term it's ok to disable the gpu on webviews. Long term it's probably not. But we won't need this once https://github.com/status-im/status-desktop/pull/16443 is merged. Found a way to overcome the freeze on 5.15.8 without disabling anything.

alexjba commented:
`export QT_XCB_GL_INTEGRATION="none"` has a few side-effects on the app built with nix (not sure yet if it's the only thing to blame):
1. Missing PNGs
<img width="198" alt="Screenshot 2024-10-07 at 14 56 26" src="https://github.com/user-attachments/assets/069a67e5-8263-4734-bfe3-8f124b93bda3">
<img width="1237" alt="Screenshot 2024-10-07 at 14 56 12" src="https://github.com/user-attachments/assets/95bde9a4-813a-4f87-8287-004b45971704">

2. Missing background and popups
<img width="738" alt="Screenshot 2024-10-07 at 14 57 46" src="https://github.com/user-attachments/assets/a75f3125-92c9-4179-aa20-fdf11dd79d61">

yakimant commented:
nixGL issue, when used with `20.09` nixpkgs: https://github.com/nix-community/nixGL/issues/183

alexjba commented:
> nixGL issue, when used with `20.09` nixpkgs: [nix-community/nixGL#183](https://github.com/nix-community/nixGL/issues/183)
> 
> bisected nixGL to find the last working commit and it works! [4469a6b](https://github.com/status-im/status-desktop/commit/4469a6b2c1b1d15dc8c8b64c7fa16257d0b4b18a)
> 
> @alexjba, please try `make run`, I will test `run-storybook` and `run-linux-gdb` meanwhile.

I'm getting

```
qt.qpa.xcb: could not connect to display
qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: eglfs, linuxfb, minimal, minimalegl, offscreen, vnc, xcb.

SIGABRT: Abnormal termination.
make[1]: *** [Makefile:864: run-linux] Aborted (core dumped)
```

alexjba commented:
> > nixGL issue, when used with `20.09` nixpkgs: [nix-community/nixGL#183](https://github.com/nix-community/nixGL/issues/183)
> > bisected nixGL to find the last working commit and it works! [4469a6b](https://github.com/status-im/status-desktop/commit/4469a6b2c1b1d15dc8c8b64c7fa16257d0b4b18a)
> > @alexjba, please try `make run`, I will test `run-storybook` and `run-linux-gdb` meanwhile.
> 
> I'm getting
> 
> ```
> qt.qpa.xcb: could not connect to display
> qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
> This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
> 
> Available platform plugins are: eglfs, linuxfb, minimal, minimalegl, offscreen, vnc, xcb.
> 
> SIGABRT: Abnormal termination.
> make[1]: *** [Makefile:864: run-linux] Aborted (core dumped)
> ```

NVM! Had an issue with the display and all graphical apps would fail to find the display.

@yakimant Works nicely!

yakimant committed:
chore(nix)_: add nix setup script

yakimant committed:
chore(nix)_: switch to nix flake

yakimant committed:
chore(nix)_: add direnv support

yakimant committed:
chore(ci)_: switch to Nix completely

yakimant committed:
chore(nix)_: fix qtwebengine issues

yakimant committed:
chore(nix)_: enforce nix shell for Linux

yakimant committed:
chore(makefile)_: add help

yakimant committed:
chore(makefile)_: switch shell to nix flake

yakimant committed:
fix

yakimant committed:
fix direnv shell

yakimant committed:
chore(nix)_: add shell welcome note

yakimant committed:
fix ;

jakubgs commented on Makefile:
```diff
@@ -104,6 +135,13 @@ else
+check-nix-shell:
+ifeq ($(detected_OS),Linux)
+ ifndef IN_NIX_SHELL
+	$(error Running outside of Nix shell is not supported)
```
Harsh :D.

jakubgs reviewed (approved):
Looks nice. Good work.

caybro commented on Makefile:
```diff
@@ -825,15 +863,15 @@ ICON_TOOL := node_modules/.bin/fileicon
 	echo -e "\033[92mRunning:\033[39m bin/nim_status_client"
-	LD_LIBRARY_PATH="$(QT5_LIBDIR)":"$(STATUSGO_LIBDIR)":"$(STATUSKEYCARDGO_LIBDIR):$(LD_LIBRARY_PATH)" \
-	./bin/nim_status_client $(ARGS)
+	nixGL ./bin/nim_status_client $(ARGS)
```
It should still be imho possible to run the binary natively (like I'm doing now); I'm not personally interested in running in a sandbox with an outdated Qt version when my system is perfectly fine

caybro replied:
I'll make a separate comment what I'd want; seems there's a general misundestanding :)

yakimant replied:
@caybro You have a different case. You build and run with same libraries.
We build with Nix libraries, running it with system libraries will cause certain issues.
The same issues, which we fix by bundling libraries into AppImage.

As I mentioned already switching 5.15.8 to 5.15.15 in out Nix setup will not be a big problem.

What is the benefit of you developing with new version btw, if we still deploy an app built with Qt 5.15.2 and bundled with it's libraries?

yakimant replied:
@alexjba I read through couple of times, but could not get it, let's talk on Discord.

But once again, you cant just run `nim_status_client` build with Nix outside of Nix shell.
It will require solving several issues.

caybro commented on Makefile:
```diff
@@ -365,7 +403,7 @@ storybook-build: | storybook-configure
 run-storybook: storybook-build
 	echo -e "\033[92mRunning:\033[39m Storybook"
-	$(STORYBOOK_BUILD_PATH)/bin/Storybook
+	nixGL $(STORYBOOK_BUILD_PATH)/bin/Storybook
```
Same here, enforcing nix is harsh...

yakimant replied:
`nixGL` is not enforcing. It's a wrapper, which fixed running an app in Nix shell with opengl hardware acceleration.
`check-nix-shell` is actually enforing Nix shell for `nim_status_client` build on Linux.

caybro replied:
Got it; will sum up my thoughts in a standalone comment

caybro commented on nix/shell.nix:
```diff
@@ -0,0 +1,108 @@
+}:
+
+let
+  qtCustom = (with pkgs.qt515_8;
```
Another point in case, the latest released Qt version is 5.15.15, why do we have to use sth that's years old?

yakimant replied:
This was originally requested by:
https://github.com/status-im/status-desktop/issues/9350

Switching to newer version will not be a problem. It's a next step.

yakimant replied:
If with switching to Qt 5.15.15 I will not need to implement and opt-out logic (to build and run with system libs) - I would certainly prefer this.

caybro replied:
Deal ;) I can test this locally on my machine next week, and provide patches/PR on top of this PR if needed

alexjba replied:
haha, nice middle ground! 😆 Can't wait for the next iteration, on windows this time.😄

yakimant commented on nix/shell.nix:
```diff
@@ -0,0 +1,108 @@
+
+let
+  qtCustom = (with pkgs.qt515_8;
+    # TODO:check the required modules
```
This list was made to build an app for AppImage.
It is build succesfully and tested by @anastasiyaig.
Maybe testing on Wayland was missing. We need to check this.

As for running locally with Wayland with `make run` - I will test in on my VM.

caybro commented:
To summarize my concerns here; the usecase is pretty simple, I as a developer want to be able to:
- build/link the application natively, that is using the OS's toolchain
- run it natively, either directly or via gdb
- be able to use these under any IDE

The compatibility changes I'd like to propose are pretty simple, touching just the `Makefile`, and not disrupting this NIX effort at all:
- introduce a new set of makefile compat targets that continue doing what they are now in master 
- the "native" compat targets would continue using the system Qt version via the existing `check-qt-dir`

For (a simple) example, the new `run-storybook` is using NIX >= add a compat target called `run-storybook-native`, and copy the old code there.

Similarly, old `run-linux` becomes `run-linux-native`, `run-linux-gdb` => `run-linux-gdb-native`, and so on with also defining `nim_status_client-native` w/o the NIX checks

jakubgs commented:
>introduce a new set of makefile compat targets that continue doing what they are now in master

I think that's a bad idea that will just introduce more complexity. We should support only one way of doing things and that's using Nix for the platforms that have been made to work with it, Linux first. If someone wants to not use it and have fun with dependency issues that's their problem and they can have fun with that.

Mobile and status-go already are essentially built using only tools provided via Nix(aside from Xcode for iOS builds), and it has improved stability of both local and CI builds. We want to achieve this in desktop as well. I can understand finding a middle ground while Nix setup is being stabilized, but ultimately we should support only one way of building things.

The only snag in this plan for Desktop is Windows platform.

jakubgs commented:
It's possible to use Nix on Windows using WSL. People do do it, it's just not something we've done before, so it will require more research and work.

yakimant commented:
@anastasiyaig, did you (and CI e2e tests) were using X11 or Wayland?

We need to make sure AppImage runs well on both.

To check (this is Wayland):
```sh
$ printnv XDG_SESSION_TYPE WAYLAND_DISPLAY
wayland
wayland-1
```

yakimant commented:
Testing `make run` on Wayland.

@micieslak reported this:
![image](https://github.com/user-attachments/assets/c644868a-094a-440d-b50c-79c6b2fa94c4)
and related issue:
https://github.com/cypress-io/cypress/issues/30160

I tried to reproduced it, but faces a different issue:
```
...
qt.qpa.plugin: Could not find the Qt platform plugin "wayland" in ""
...
WRN 2024-10-16 10:36:08.352Z qt warning                                 topics="qt" tid=71966 text="QQmlApplicationEngine failed to load component" file=:0 category=default                    
WRN 2024-10-16 10:36:08.359Z qt warning                                 topics="qt" tid=71966 text="module \"StatusQ.Core.Theme\" is not installed" file=qrc:/main.qml:20 category=default      
...                                                                                   
WRN 2024-10-16 10:36:08.412Z qt warning                                 topics="qt" tid=71966 text="Error while loading QML: QUrl(\"qrc:/main.qml\")" file=:0 category=default                  
...
ERROR[10-16|10:36:08.418|github.com/status-im/status-go/mobile/status.go:865] error in makeJSONResponse                error="there is no running node"
```

yakimant committed:
fix wayland + gnome issue

yakimant commented:
Tests don't work well:
```sh
$ make run-storybook-tests
make[1]: Entering directory '/home/s/work/status-desktop'
Building: Storybook
gmake[2]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
Running: Storybook Tests
UpdateCTestConfiguration  from :/home/s/work/status-desktop/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/s/work/status-desktop/DartConfiguration.tcl
Test project /home/s/work/status-desktop
Constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
No tests were found!!!
make[1]: Leaving directory '/home/s/work/status-desktop'
```

```sh
$ make run-statusq-tests
make[1]: Entering directory '/home/s/work/status-desktop'
Configuring: StatusQ Unit Tests
Building: StatusQ Unit Tests
gmake[2]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
Running: StatusQ Unit Tests
UpdateCTestConfiguration  from :/home/s/work/status-desktop/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/s/work/status-desktop/DartConfiguration.tcl
Test project /home/s/work/status-desktop
Constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
No tests were found!!!
make[1]: Leaving directory '/home/s/work/status-desktop'
```

yakimant commented:
Failing test:
```
test 1
      Start  1: QmlTests

1: Test command: /home/s/work/status-desktop/ui/StatusQ/build/bin/QmlTests "-input" "/home/s/work/status-desktop/ui/StatusQ/tests"
1: Test timeout computed to be: 10000000
1: Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.
1: 
1: (process:146863): Gtk-WARNING **: 16:08:50.562: Locale not supported by C library.
1:      Using the fallback 'C' locale.
1: Gtk-Message: 16:08:50.883: Failed to load module "canberra-gtk-module"
1: Gtk-Message: 16:08:50.886: Failed to load module "canberra-gtk-module"
1: qt.glx: qglx_findConfig: Failed to finding matching FBConfig for QSurfaceFormat(version 2.0, options QFlags<QSurfaceFormat::FormatOption>(), depthBufferSize -1, redBufferSize 1, greenBufferSize 1, blueBufferSize 1, alphaBufferSize -1, stencilBufferSize -1, samples -1, swapBehavior QSurfaceFormat::SingleBuffer, swapInterval 1, colorSpace QSurfaceFormat::DefaultColorSpace, profile  QSurfaceFormat::NoProfile)
1: qt.glx: qglx_findConfig: Failed to finding matching FBConfig for QSurfaceFormat(version 2.0, options QFlags<QSurfaceFormat::FormatOption>(), depthBufferSize -1, redBufferSize 1, greenBufferSize 1, blueBufferSize 1, alphaBufferSize -1, stencilBufferSize -1, samples -1, swapBehavior QSurfaceFormat::SingleBuffer, swapInterval 1, colorSpace QSurfaceFormat::DefaultColorSpace, profile  QSurfaceFormat::NoProfile)
1: Could not initialize GLX
 1/16 Test  #1: QmlTests .........................Subprocess aborted***Exception:   1.72 sec
```

yakimant committed:
fix ui tests

yakimant committed:
fix jenkins tests

yakimant commented:
@micieslak reported another issue:
```
Building: DOtherSide
gmake[2]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
In file included from /home/michalc/workspace/status-desktop/vendor/qzxing/src/zxing/zxing/ResultIO.cpp:22:
/home/michalc/workspace/status-desktop/vendor/qzxing/src/zxing/zxing/Result.h:57:24: note: only here as a ‘friend’
   57 |   friend std::ostream& operator<<(std::ostream &out, Result& result);
      |                        ^~~~~~~~
/nix/store/vfqlryhvm8063hs7ax9k2vb8wmch5v0v-binutils-2.31.1/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/11/liblto_plugin.so: error loading plugin: /nix/store/0c7c96gikmzv87i7lv3vq5s1cmfjd6zf-glibc-2.31-74/lib/libc.so.6: version `GLIBC_2.33' not found (required by /usr/lib/gcc/x86_64-linux-gnu/11/liblto_plugin.so)
collect2: error: ld returned 1 exit status
```

looks like non-nix and nix build were mixed. `make clean && make clean-git` helped.
